### ISSUE TITLE
[Performance] AscendSampler vs v2 Gumbel Sampling: ~55× latency gap on NPU

### ISSUE THREAD

**Issue by realliujiaxu** [vllm-project/vllm-ascend#7608]
## Summary

Benchmark shows that the **v1 AscendSampler** (`vllm_ascend/sample/sampler.py`) sampling path is **~56× faster** than the **v2 Gumbel Triton kernel** (`vllm_ascend/worker/v2/sample/gumbel.py`) under the same workload. Both paths implement Gumbel-max categorical sampling; the gap suggests optimization opportunity for v2 model runner.

## Environment

- **Device**: Ascend NPU
- **Timing**: `torch.npu.Event` (NPU-side only)

## Two Sampling Paths

| Path | Location | Implementation |
|------|----------|----------------|
| **AscendSampler (v1)** | `vllm_ascend/sample/sampler.py` | `softmax` → `exponential` → `probs.div_(q)` → `argmax` (PyTorch) |
| **v2 Gumbel** | `vllm_ascend/worker/v2/sample/gumbel.py` | Triton kernel: add Gumbel noise to logits block-wise, then argmax per block and across blocks |

Both are mathematically equivalent (Gumbel-max trick for categorical sampling).

## Benchmark Results

**Config**: `vocab_size=128000`, `temperature=0.8`, `num_reqs ∈ {1, 8, 32, 128}`, warmup=5, iters=50

| num_reqs | AscendSampler (sample/sampler.py) | v2 gumbel.py |
|----------|-----------------------------------|--------------|
| 1 | 0.54 ms | 1.40 ms |
| 8 | 0.56 ms | 7.90 ms |
| 32 | 0.89 ms | 31.12 ms |
| 128 | 2.25 ms | 124.00 ms |

**Speedup (AscendSampler / v2 gumbel)**: ~2.6× (num_reqs=1) → **~55×** (num_reqs=128)

## Reproducibility

Copy the following script to a `.py` file and run it in the vllm-ascend repo (with NPU available):

```python
#!/usr/bin/env python3
"""AscendSampler vs v2 gumbel performance comparison.
Uses AscendSampler instance and forward().
"""
import torch

from vllm.config import VllmConfig
from vllm.v1.sample.logits_processor import LogitsProcessors
from vllm.v1.sample.metadata import SamplingMetadata
from vllm_ascend.ascend_config import init_ascend_config
from vllm_ascend.ops.triton.triton_utils import init_device_properties_triton
from vllm_ascend.sample.sampler import AscendSampler
from vllm_ascend.worker.v2.sample.gumbel import gumbel_sample

VOCAB_SIZE = 128000
NUM_REQS_LIST = [1, 8, 32, 128]
NUM_WARMUP = 5
NUM_ITERS = 50


def make_sampling_metadata(num_reqs: int, temperature: torch.Tensor, device: str):
    return SamplingMetadata(
        temperature=temperature,
        all_greedy=False,
        all_random=True,
        top_p=None,
        top_k=None,
        generators={},
        max_num_logprobs=None,
        no_penalties=True,
        prompt_token_ids=None,
        frequency_penalties=torch.zeros(num_reqs, dtype=torch.float32, device=device),
        presence_penalties=torch.zeros(num_reqs, dtype=torch.float32, device=device),
        repetition_penalties=torch.ones(num_reqs, dtype=torch.float32, device=device),
        output_token_ids=[[] for _ in range(num_reqs)],
        allowed_token_ids_mask=None,
        bad_words_token_ids={},
        logitsprocs=LogitsProcessors(),
    )


def main():
    if not torch.npu.is_available():
        print("SKIP: NPU not available")
        return

    init_device_properties_triton()
    init_ascend_config(VllmConfig())
    device = "npu"
    torch.manual_seed(42)

    start_evt = torch.npu.Event(enable_timing=True)
    end_evt = torch.npu.Event(enable_timing=True)
    ascend_sampler = AscendSampler()

    print("AscendSampler vs v2 gumbel performance comparison")
    print("=" * 60)
    print(f"vocab_size={VOCAB_SIZE}, warmup={NUM_WARMUP}, iters={NUM_ITERS}")
    print()

    for num_reqs in NUM_REQS_LIST:
        logits = torch.randn(num_reqs, VOCAB_SIZE, dtype=torch.float32, device=device)
        temperature = torch.ones(num_reqs, dtype=torch.float32, device=device) * 0.8
        idx_mapping = torch.arange(num_reqs, dtype=torch.int64, device=device)
        seed = torch.randint(0, 2**31, (num_reqs,), dtype=torch.int64, device=device)
        pos = torch.randint(0, 100, (num_reqs,), dtype=torch.int64, device=device)
        sampling_metadata = make_sampling_metadata(num_reqs, temperature, device)

        def run_ascend_sampler():
            return ascend_sampler.forward(logits.clone(), sampling_metadata)

        for _ in range(NUM_WARMUP):
            run_ascend_sampler()
        torch.npu.synchronize()
        times = []
        for _ in range(NUM_ITERS):
            start_evt.record()
            run_ascend_sampler()
            end_evt.record()
            torch.npu.synchronize()
            times.append(start_evt.elapsed_time(end_evt))
        t_ascend = sum(times) / NUM_ITERS

        def run_gumbel():
            gumbel_sample(logits, idx_mapping, temperature, seed, pos, apply_temperature=True)

        for _ in range(NUM_WARMUP):
            run_gumbel()
        torch.npu.synchronize()
        times = []
        for _ in range(NUM_ITERS):
            start_evt.record()
            run_gumbel()
            end_evt.record()
            torch.npu.synchronize()
            times.append(start_evt.elapsed_time(end_evt))
        t_gumbel = sum(times) / NUM_ITERS

        speedup = t_gumbel / t_ascend if t_ascend > 0 else 0
        print(f"num_reqs={num_reqs}: AscendSampler {t_ascend:.3f} ms | v2 gumbel {t_gumbel:.3f} ms | speedup {speedup:.1f}x")

    print("=" * 60)


if __name__ == "__main__":
    main()
```

## Related

- v2 Model Runner: [Model Runner V2](https://github.com/vllm-project/vllm-ascend/issues/5208)
- AscendSampler: `vllm_ascend/sample/sampler.py`
- v2 Gumbel: `vllm_ascend/worker/v2/sample/gumbel.py`

**Comment by realliujiaxu** [comment]
@Ronald1995 Could we track this performance issue in model runner v2 RFC?
